src/Features/Core/Portable/CodeRefactorings/AddMissingImports/AbstractAddMissingImportsFeatureService.cs

```diff
@@ -0,0 +1,150 @@
+                    if (fixesForDiagnostic.Fixes.Length == 1)
+                    {
+                        fixes.Add(fixesForDiagnostic.Fixes[0]);
+                    }
+                }
+            }
+
+            return fixes.ToImmutableAndFree();
+        }
+
+        private async Task<Document> ApplyFixesAsync(Document document, ImmutableArray<AddImportFixData> fixes, CancellationToken cancellationToken)
+        {
+            if (fixes.IsEmpty)
+            {
+                return document;
+            }
+
+            var solution = document.Project.Solution;
+            var packageInstallerService = solution.Workspace.Services.GetService<IPackageInstallerService>();
+            var addImportService = document.GetLanguageService<IAddImportFeatureService>();
+
+            // Do not limit the results since we plan to fix all the reported issues.
+            var codeActions = addImportService.GetCodeActionsForFixes(document, fixes, packageInstallerService, limitResults: false);
+            var getChangesTasks = codeActions.Select(action => GetChangesForCodeActionAsync(document, action, cancellationToken));
+
+            await Task.WhenAll(getChangesTasks).ConfigureAwait(false);
+
+            // Using Sets allows us to accumulate only the distinct changes.
+            var allTextChanges = new HashSet<TextChange>();
+            // Some fixes require adding missing references.
+            var allAddedProjectReferences = new HashSet<ProjectReference>();
+            var allAddedMetaDataReferences = new HashSet<MetadataReference>();
+
+            foreach (var getChangesTask in getChangesTasks)
+            {
+                var (projectChanges, textChanges) = await getChangesTask.ConfigureAwait(false);
+
+                allTextChanges.UnionWith(textChanges);
+                allAddedProjectReferences.UnionWith(projectChanges.GetAddedProjectReferences());
+                allAddedMetaDataReferences.UnionWith(projectChanges.GetAddedMetadataReferences());
+            }
+
+            // Apply changes to both the project and document.
+            var newProject = document.Project.AddMetadataReferences(allAddedMetaDataReferences);
+            newProject = newProject.AddProjectReferences(allAddedProjectReferences);
+
+            var text = await document.GetTextAsync(cancellationToken);
+            var newText = text.WithChanges(allTextChanges);
+            var newDocument = newProject.GetDocument(document.Id).WithText(newText);
+
+            return newDocument;
+        }
+
+        private async Task<(ProjectChanges projectChanges, IEnumerable<TextChange> textChanges)> GetChangesForCodeActionAsync(Document document, CodeAction codeAction, CancellationToken cancellationToken)
+        {
+            var newSolution = await codeAction.GetChangedSolutionInternalAsync(cancellationToken: cancellationToken).ConfigureAwait(false);
```

Comment:
hrmm... why calling this internal method?

Reply:
The non-internal method takes a ProgressTracker which the callers into this method don't have. I thought I was saving the cost of instantiating a new ProgressTracker, but I now see the Internal method creates a new one anyway. I'll switch back.

Reply:
Bleagh.  Something is really wrong with all the progress tracking code in hte IDE currently.  We should be passing along the real progress-tracker that is used for doing progress updates in the dialog while the fix applies.  but that's definitely not your fault here.  so don't worry about it too much.  i'll have to see if i can fix this all up at some point :(